sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala

```diff
@@ -94,6 +94,7 @@ object ReplaceNullWithFalseInPredicate extends Rule[LogicalPlan] {
         replaceNullWithFalse(cond) -> replaceNullWithFalse(value)
       }
       val newElseValue = cw.elseValue.map(replaceNullWithFalse)
+        .orElse(if (newBranches.forall(_._2 == FalseLiteral)) Some(FalseLiteral) else None)
```

Comment:
This is a weird change. Can we update `SimplifyConditionals` to eliminate CASE WHEN branches even without the else value?

Reply:
It seems it should be in `ReplaceNullWithFalseInPredicate`, this is an example in Project:
```
spark-sql> create table t1 using parquet as select id from range(5);
spark-sql> select (CASE WHEN id = 1 THEN 'a' WHEN id = 3 THEN 'b' end) = 'c' from t1;
NULL
false
false
NULL
NULL
```

```
spark-sql> select * from t1 where (CASE WHEN id = 1 THEN 'a' WHEN id = 3 THEN 'b' end) = 'c';
spark-sql>
```

Reply:
ah now I got it. We can probably eliminate CASE WHEN branches without ELSE if the branch value is null literal. But that's a different topic.

Reply:
I think one model can be:
1. CASE WHEN always has an ELSE branch. If not specified, it's null literal.
2. `ReplaceNullWithFalseInPredicate` and `SimplifyConditionals` should both consider the no-ELSE case (which means null literal).

Reply:
This will change this behaviour:
https://github.com/apache/spark/pull/30852/files#diff-906b229035e0b7e8a0148ed6342d24a83f7afc769891a472df248b9a3ff9aa10R137-R138

Reply:
Can you be specific about this behavior change? Does a query change its result?

Reply:
How Another just add this to `SimplifyConditionals`? The logic is clear and simple:
```scala
case cw @ CaseWhen(_, elseValue) if cw.dataType == BooleanType && elseValue.isEmpty =>
  cw.copy(elseValue = Some(FalseLiteral))
```

Reply:
sorry, it is incorrect.

Reply:
The result will not change. We will not push foldable into branches for `EqualTo(CaseWhen(Seq((a, b)), None), Literal(1))` before this pr, but we will push foldable into branches after this pr, because we will rewrite it to `EqualTo(CaseWhen(Seq((a, b)), null), Literal(1))`.

Reply:
I didn't mean we should blindly fill the ELSE branch with null literal. We can virtually treat unspecified ELSE as null literal, to enhance the logic in `SimplifyConditionals` and `ReplaceNullWithFalseInPredicate`.

For example, `SimplifyConditionals` can do something like
```
case e @ CaseWhen(branches, elseOpt)
    if branches.forall(_._2.semanticEquals(elseOpt.getOrElse(Literal(null, e.dataType)))) =>
  ...
```
It won't change the expression if there is no benefit.

Reply:
Changed it to:
```scala
case e @ CaseWhen(branches, elseOpt)
    if elseOpt.isEmpty && branches.forall(_._2.semanticEquals(Literal(null, e.dataType))) =>
  Literal(null, e.dataType)
```
This is only handle elseOpt.isEmpty, elseOpt.nonEmpty has handled by SPARK-24893.